handlers/inmem/inmem.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2016 Netflix, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inmem
+
+import (
+	"sync"
+	"time"
+
+	"github.com/netflix/rend/common"
+	"github.com/netflix/rend/handlers"
+)
+
+type entry struct {
+	data    []byte
+	instime int64
+	exptime int64
+	flags   uint32
+}
+
+type Handler struct {
+	data  map[string]entry
+	mutex *sync.RWMutex
+}
+
+var singleton = &Handler{
+	data:  make(map[string]entry),
+	mutex: new(sync.RWMutex),
+}
+
+func New() (handlers.Handler, error) {
+	// return the same singleton map each time so all connections see the same data
+	return singleton, nil
+}
+
+func (h *Handler) Set(cmd common.SetRequest) error {
+	h.mutex.Lock()
+
+	key := string(cmd.Key)
+	instime := time.Now().Unix()
+	exptime := instime + int64(cmd.Exptime)
+	h.data[key] = entry{
+		data:    cmd.Data,
+		instime: instime,
+		exptime: exptime,
+		flags:   cmd.Flags,
+	}
+
+	h.mutex.Unlock()
+	return nil
+}
+
+func (h *Handler) Add(cmd common.SetRequest) error {
+	h.mutex.Lock()
+
+	key := string(cmd.Key)
+
+	if _, ok := h.data[key]; ok {
+		h.mutex.Unlock()
+		return common.ErrKeyExists
+	}
+
+	instime := time.Now().Unix()
+	exptime := instime + int64(cmd.Exptime)
+
+	h.data[key] = entry{
+		data:    cmd.Data,
+		instime: instime,
+		exptime: exptime,
+		flags:   cmd.Flags,
+	}
+
+	h.mutex.Unlock()
+	return nil
+}
+
+func (h *Handler) Replace(cmd common.SetRequest) error {
+	h.mutex.Lock()
+
+	key := string(cmd.Key)
+
+	if _, ok := h.data[key]; !ok {
+		h.mutex.Unlock()
+		return common.ErrKeyExists
+	}
+
+	instime := time.Now().Unix()
+	exptime := instime + int64(cmd.Exptime)
+
+	h.data[key] = entry{
+		data:    cmd.Data,
+		instime: instime,
+		exptime: exptime,
+		flags:   cmd.Flags,
+	}
+
+	h.mutex.Unlock()
+	return nil
+}
+
+func (h *Handler) Get(cmd common.GetRequest) (<-chan common.GetResponse, <-chan error) {
+	dataOut := make(chan common.GetResponse, len(cmd.Keys))
+	errorOut := make(chan error)
+
+	h.mutex.RLock()
+
+	for idx, bk := range cmd.Keys {
+		key := string(bk)
+		e, ok := h.data[key]
```

Comment:
use `e, ok  := h.data[string(bk)]`

The Go compiler has an optimization that eliminates the slice->string copy.